### PR TITLE
customizable registries

### PR DESCRIPTION
**Pull request by Glease**
## features

1. Allow a player to configure stack extractors and inventory providers. These can now be reordered and separately disabled.
  * stack extractors extract inventories from items, e.g. connect to ME network linked with a wireless me inventory
  * inventory providers extract inventories from players, e.g. main 36 slots player inventory
2. Configurations can be modified either when the game is closed, or when the player is at main menu, or when player is in game (*).
3. Configurations will be sent to server upon server join or reconfigure. Each player will uses his own copy of configuration he defined locally. If he doesn't have one, the one configured on server side will take effect.

*: require EnderCore, which already comes with the pack.
## demo

See [#upcoming-features on discord](https://discord.com/channels/181078474394566657/295669878222880769/1232702597945626735).

## todo
- [x] per player configuration
- [x] network sync
- [x] config gui

### CODE REVIEW

**Comment by Glease** [comment]
this is now ready to be reviewed.

File: src/main/java/com/gtnewhorizon/structurelib/GuiEditRegistryConfig.java [hunk 1]
```diff
@@ -0,0 +1,702 @@
...
+        }
+
+        this.btnUndoAll.enabled = isChanged();
+        this.btnDefault.enabled = isDefault();
+        super.drawScreen(mouseX, mouseY, partialTicks);
+        for (EntriesList list : lists) {
+            list.drawScreenPost(mouseX, mouseY, partialTicks);
+        }
+        if (this.undoHoverChecker.checkHover(mouseX, mouseY)) this.drawToolTip(
+                this.mc.fontRenderer.listFormattedStringToWidth(I18n.format("fml.configgui.tooltip.undoAll"), 300),
```
Comment (Quarri6343) [inline src/main/java/com/gtnewhorizon/structurelib/GuiEditRegistryConfig.java:323]: listFormattedStringToWidth does not consider the distance between mouse position and the screen edge, so the auto line break does not work at all.
I, the Auto Gui Scale enjoyer, can't read most of tooltips in the config screen properly.
スクリーンショット 2024-05-05 162751

File: src/main/java/com/gtnewhorizon/structurelib/item/ItemConstructableTrigger.java [hunk 2]
```diff
@@ -28,8 +28,13 @@ public ItemConstructableTrigger() {
 
     @Override
     public ItemStack onItemRightClick(ItemStack stack, World world, EntityPlayer player) {
-        if (world.isRemote && getMovingObjectPositionFromPlayer(world, player, true) == null)
-            player.openGui(StructureLib.instance(), 0, world, player.inventory.currentItem, 0, 0);
+        if (world.isRemote && getMovingObjectPositionFromPlayer(world, player, true) == null) {
+            if (player.isSneaking()) {
```
Comment (Quarri6343) [inline src/main/java/com/gtnewhorizon/structurelib/item/ItemConstructableTrigger.java:32]: If you don't mind, please add a line to projector to tell people sneak use on air to open the new gui.

File: src/main/java/com/gtnewhorizon/structurelib/GuiEditRegistryConfig.java [hunk 3]
```diff
@@ -0,0 +1,702 @@
...
+                        2001,
+                        this.width / 2 - buttonWidthHalf + doneWidth + 5 + undoWidth + 5,
+                        this.height - 29,
+                        resetWidth,
+                        20,
+                        " " + I18n.format("fml.configgui.tooltip.resetToDefault"),
+                        RESET_CHAR,
+                        2.0F));
+        this.buttonList.add(
+                btnUndoAll = new GuiUnicodeGlyphButton(
```
Comment (Quarri6343) [inline src/main/java/com/gtnewhorizon/structurelib/GuiEditRegistryConfig.java:169]: Reset to default button messes up if you click done button in the top config menu
(if you press escape twice the default setting will not be overridden)

https://github.com/GTNewHorizons/StructureLib/assets/39122497/a7c85fdc-2d01-4375-87c0-14ef67d597aa

**Review by Quarri6343 — COMMENTED** [review]
・Stack Extractor does not work when Inventory Provider is disabled.
If this is an intentional behavior, I suggest blocking players from opening Stack Extractor config screen when Inventory Provider is disabled.

https://github.com/GTNewHorizons/StructureLib/assets/39122497/58ccb996-3e3f-468c-9c99-8f4979b587ff

・The projector says blocks are built even if the wireless terminal runs out of power

スクリーンショット 2024-05-05 180646

**Comment by Glease** [comment]
> Stack Extractor does not work when Inventory Provider is disabled.

Stack extractors operate on the stacks provided by inventory provider. Tweaking stack extractors is indeed meaningless if all of inventory provider is disabled. However I don't see why player should be blocked from tweaking stack extractors. Perhaps a warning is what we all need.

> projector says blocks are built even if the wireless terminal runs out of power

Maybe structurecompat was broken in some iteration of the wireless terminals. needs to be fixed separately there.

**Comment by Quarri6343** [comment]
> Perhaps a warning is what we all need.

Warning is fine as long as people don't post the issue about it

**Comment by mitchej123** [comment]
@Glease Do you have time to address the issues raised?

@Quarri6343 Are any of these show stoppers?  

It looks like Dream's been shipping this on Zeta for a while, so probably fine to merge, depending on feedback from above

**Comment by Pxx500** [comment]
@Glease this has been up very long and it's a massive QoL, are you planning on finishing it or maybe someone else can finish it?

**Comment by Dream-Master** [comment]
I pm Glease already but get no answer so far. Maybe he is very busy?

**Review by Dream-Master — APPROVED** [review]

**Pull request merged**